プレハブの検査機能を追加

ku6ryo opened the pull request:
1. Project Windowのプレハブを選択して、コンテキストメニューから「STYLY/Validate Prefab」を実行
2. 検証項目は PrefabValidatorRunner内に記載
3. 現時点で検証できる機能は以下の通り - 利用不可能なシェーダーを使用していないか？ - 利用不可能なコンポーネントを使用していないか？
4. 検査結果はConsoleログに出力

zabaglione committed:
プレハブの検査機能を追加
1. Project Windowのプレハブを選択して、コンテキストメニューから「STYLY/Validate Prefab」を実行
2. 検証項目は PrefabValidatorRunner内に記載
3. 現時点で検証できる機能は以下の通り
     - 利用不可能なシェーダーを使用していないか？
     - 利用不可能なコンポーネントを使用していないか？
4. 検査結果はConsoleログに出力

ku6ryo commented on Assets/Validation/Editor/Validators/ShaderValidator.cs:
```diff
@@ -0,0 +1,46 @@
+            return passed;
+        }
+
+        private string GetGameObjectPath(GameObject obj)
```
これはどのクラスでも使いそうなので、どこかで共通で使えるようにしてもらえますでしょか？

zabaglione replied:
Utilityクラスを作成して共有化しました。

ku6ryo commented on Assets/Validation/Editor/Validators/ShaderValidator.cs:
```diff
@@ -0,0 +1,46 @@
+            string path = obj.name;
+            while (obj.transform.parent != null)
+            {
+                obj = obj.transform.parent.gameObject;
```
引数を上書きするとバグが実装されてしまう確率があがるので、別の変数に持ち替えて実装してください。

zabaglione replied:
作業用変数を用意しました。

ku6ryo reviewed (commented):
基本的な設計はいいと思います細かい点コメントしました。

@afjk さんにもレビューを依頼します。

ku6ryo commented:
メニューのラベルについて
「Validate Prefab (Alpha)」として置いてもらえますでしょうか？

動作について、Standard shader を使ったマテリアルの obj を含めましたが、検査に引っ掛かりませんでした。動作バグがあるかもです。
![スクリーンショット 2024-09-10 015321](https://github.com/user-attachments/assets/7e9ae48a-0c08-4d8a-99b5-be0a300a562a)

ku6ryo commented on Assets/Validation/Editor/Validators/ShaderValidator.cs:
```diff
@@ -0,0 +1,46 @@
+                    if (material != null && System.Array.IndexOf(allowedShaders, material.shader) == -1)
+                    {
+                        string path = GetGameObjectPath(renderer.gameObject);
+                        Debug.LogWarning($"Using unauthorized shaders: {material.shader.name}, Game Object: {path}");
```
Unauthorized だと、ログイン的な認証されていないものという意味になるので、unsupported に変更できますでしょうか？

zabaglione replied:
変更します。

zabaglione replied:
メニュー名は「Validate Prefab (Alpha)」に変更します。

zabaglione committed:
- プレハブ内のパス名の生成関数をUtilityクラスで共通化
- メニュー名を「Validate Prefab (Alpha)」に変更
- プレハブ内のパス名の生成関数内の処理において、引数の値を直接変更せずに作業用変数を用意して変更する

zabaglione committed:
TextMesh Proのシェーダーを１つ「有効シェーダー」として追加

afjk commented on Assets/Validation/Editor/Utility/ValidatorUtility.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Styly.VisionOs.Plugin.Validation
+{
+    public class ValidatorUtility : MonoBehaviour
```
MonoBehaviourの継承は不要かと思います。
削除お願いします。

zabaglione replied:
最新コミットで削除しました。

afjk replied:
ありがとうございます！

afjk commented:
パスについて
/Packages/com.styly.styly-spatial-layer-plugin
がPluginとしての配布パッケージになりますので、Validation機能もこちらに含めていただくようお願いします。

afjk commented on Assets/Validation/Editor/Main/PrefabValidatorRunner.cs:
```diff
@@ -0,0 +1,94 @@
+            PrefabValidationManager validationManager = new PrefabValidationManager();
+
+            // Added shader verification
+            Shader[] allowedShaders = { 
```
対応シェーダー一覧は変更される可能性が高いので外部ファイルにしていただけると助かります。

zabaglione replied:
色々と検討しましたが、Configフォルダーを作成して、そちらに機能別にクラスを分けて対応しました。ほかの方法としては、ScriptableObjectにする方法が考えられますが、パッケージ内部にあるScriptableObjectをうまく指定する方法が思いつきませんでした・・・

afjk commented on Assets/Validation/Editor/Main/PrefabValidatorRunner.cs:
```diff
@@ -0,0 +1,94 @@
+            validationManager.AddValidator(new ShaderValidator(allowedShaders));
+
+            // Added forbidden components
+            System.Type[] forbiddenComponents = {
```
非対応コンポーネント一覧は頻繁に変更する可能性が高いので外部ファイルにしていただけると助かります。

afjk commented:
> 検査結果はConsoleログに出力

結果がConsoleに出力されるだけですと、ユーザーは何が起きたのか気づきにくいので成功/失敗ダイアログで通知するなどした方が良いと思います。

afjk commented:
@zabaglione ご対応ありがとうございます。
数点コメントさせていただきました。
ご確認お願いいたします。

zabaglione committed:
・バウンディングボックスの検証コードを追加
・検証用の情報をConfigフォルダ内に移動
・プライベートフィールドの先頭に"_"を付与

zabaglione committed:
・頂点数の検査機能を追加（個別とトータルの検査）

zabaglione committed:
・テクスチャサイズの検査機能を追加（暫定で4096x4096のいずれかの一辺を超えたものはNGとする（4096x4096はOK））

zabaglione committed:
・床下チェック機能を追加（Transform.position.y < 0 のゲームオブジェクト）

zabaglione committed:
- 全ての検査終了時のメッセージを修正（Pass/Failedではなく、あくまでも検査をしたということと、実施した結果警告が出たということがわかるメッセージ）
- マテリアル関連(Shaderとテクスチャサイズ)の検査を一時的に無効化

zabaglione committed:
プレハブの検査を実行終了時にダイアログを表示

zabaglione committed:
ダイアログのタイトルを「Validate Prefab (Alpha)」に変更

zabaglione committed:
MonoBehaviourの継承は不要だったので削除

ku6ryo commented:
動作のチェックをしました。

不具合、修正いただきたい点
** Prefab 内のコンポーネントがすべてチェックされていなそう **
コードを拝見して、Prefab 内の Child の Child など深くまでチェックされていなそうにみえます。（間違っていたら教えてください）

**Warning が出そうな prefab を作っても warning がでない。**
具体的にチェックした箇所は
- GoundValidator で Position Y の値が負のオブジェクトを作成したが、Warning がでなかった。
- Camera component がついているコンポーネントを含めたが、Warning がでなかった。

上記の通り深くまでチェックされていないと思ったので、prefab の GameObject 直下のコンポーネントで確認をしています。

**Prefab を選択していないときに「Validate Prefab (Alpha)」が選択できないようにしてほしい。**
Build Prefab のメニューの設定コードを参照ください

zabaglione committed:
- プレハブ以外を選択しているときはメニューを実行できないように修正
- このクラスもMonoBehaviourの継承は不要なので削除

ku6ryo commented:
Validator 使用しました (mac)

修正していただきたい点
- ログが埋もれるので、[STYLY Validator] のようなフィルタリングできる文字列を入れる必要があ理想なので追加して欲しい
- 非推奨のコンポーネントが複数の GameObject についているときに一つしか指摘されない
- ユーザーが Camera コンポーネントを使用してシーン内のオブジェクトを撮影、RenderTexture でマテリアルに設定するみたいなケースがそこそこあるみたいです。なので Camera は対象から外してください。

zabaglione committed:
- コンポーネントの検査において、複数箇所で使用不可能のコンポーネントが使われていた時に一箇所しか検出しなかった不具合を修正
- ログメッセージに"[STYLY Validator]"のプレフィックスをつけてログを出力
- Config系クラスがMonoBehaviourからの継承になっていたけど、不要なので削除

zabaglione committed:
- Prefab に依存している Prefab のチェック機能を追加

zabaglione committed:
指定されたPrefabや再帰的に見つけたPrefabにおいて、オーバーライドされていないときは警告ダイアログを表示し、処理を継続するかキャンセルするかを指定できるようにした
